ra/registration-authority.go

```diff
@@ -146,32 +166,89 @@ func (ra *RegistrationAuthorityImpl) NewCertificate(req core.CertificateRequest,
 			continue
 		}
 
+		for _, challenge := range authz.Challenges {
+			if challenge.Status == core.StatusValid {
+				verificationMethodSet[challenge.Type] = true
+			}
+		}
+
 		authorizedDomains[authz.Identifier.Value] = true
 	}
+	verificationMethods := []string{}
+	for method, _ := range verificationMethodSet {
+		verificationMethods = append(verificationMethods, method)
+	}
 
 	// Validate that authorization key is authorized for all domains
 	names := csr.DNSNames
 	if len(csr.Subject.CommonName) > 0 {
 		names = append(names, csr.Subject.CommonName)
 	}
+
+	// Construct the log event
+	logEvent := certificateRequestEvent{
+		ID:                  core.NewToken(),
+		Requester:           registration.ID,
+		RequestMethod:       "online",
+		VerificationMethods: verificationMethods,
+		VerifiedFields:      []string{"subject.commonName", "subjectAltName"},
+		CommonName:          names[0],
+		Names:               names,
+		RequestTime:         time.Now(),
+	}
+
+	// Validate all domains
 	for _, name := range names {
 		if !authorizedDomains[name] {
 			err = core.UnauthorizedError(fmt.Sprintf("Key not authorized for name %s", name))
+
+			logEvent.Error = err.Error()
+			jsonLogEvent, logErr := json.Marshal(logEvent)
+			if logErr != nil {
```

Comment:
If we get a certificate request but fail to log it because of a JSON marshaling error, we should log that as an audit event (which won't include the JSON certificateRequestEvent, of course).